File: backend/BankingApp.Infrastructure/Services/NotificationService.cs
using BankingApp.Application.Common.Exceptions;
using BankingApp.Application.Common.Pagination;
using BankingApp.Application.Interfaces;
using BankingApp.Application.Notifications;
using BankingApp.Domain.Entities;
using BankingApp.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;

namespace BankingApp.Infrastructure.Services;

public sealed class NotificationService(BankingAppDbContext db, ICurrentUserService currentUser) : INotificationService
{
    public async Task<PagedResult<NotificationResponse>> GetAsync(NotificationQuery request, CancellationToken token = default)
    {
        var query = db.Notifications.AsNoTracking().Where(x => x.UserId == currentUser.UserId);
        if (request.UnreadOnly) query = query.Where(x => !x.IsRead);
        var total = await query.CountAsync(token);
        var entities = await query.OrderByDescending(x => x.CreatedAtUtc).Skip((request.Page - 1) * request.PageSize).Take(request.PageSize).ToListAsync(token);
        var items = entities.Select(Map).ToList();
        return new PagedResult<NotificationResponse> { Items = items, Page = request.Page, PageSize = request.PageSize, TotalCount = total };
    }
    public Task<int> GetUnreadCountAsync(CancellationToken token = default) => db.Notifications.CountAsync(x => x.UserId == currentUser.UserId && !x.IsRead, token);
    public async Task MarkReadAsync(Guid id, CancellationToken token = default)
    {
        var value = await db.Notifications.SingleOrDefaultAsync(x => x.Id == id && x.UserId == currentUser.UserId, token) ?? throw new NotFoundException("Notification was not found.");
        if (!value.IsRead) { value.IsRead = true; value.ReadAtUtc = DateTime.UtcNow; await db.SaveChangesAsync(token); }
    }
    public async Task MarkAllReadAsync(CancellationToken token = default)
    {
        var now = DateTime.UtcNow;
        if (db.Database.IsRelational())
        {
            await db.Notifications.Where(x => x.UserId == currentUser.UserId && !x.IsRead).ExecuteUpdateAsync(setters => setters.SetProperty(x => x.IsRead, true).SetProperty(x => x.ReadAtUtc, now), token);
            return;
        }
        var values = await db.Notifications.Where(x => x.UserId == currentUser.UserId && !x.IsRead).ToListAsync(token);
        foreach (var value in values) { value.IsRead = true; value.ReadAtUtc = now; }
        await db.SaveChangesAsync(token);
    }
    private static NotificationResponse Map(Notification x) => new() { Id = x.Id, Type = x.Type.ToString(), Title = x.Title, Message = x.Message, EntityType = x.EntityType, EntityId = x.EntityId, IsRead = x.IsRead, CreatedAtUtc = x.CreatedAtUtc, ReadAtUtc = x.ReadAtUtc };
}

File: ui/desktop_app/lib/src/features/notifications/notification_service.dart
import '../../core/api_client.dart';

class AdminNotification {
  const AdminNotification({
    required this.id,
    required this.type,
    required this.title,
    required this.message,
    required this.isRead,
    required this.createdAtUtc,
    this.entityType,
    this.entityId,
  });
  final String id, type, title, message;
  final String? entityType, entityId;
  final bool isRead;
  final DateTime createdAtUtc;
  factory AdminNotification.fromJson(Map<String, dynamic> json) =>
      AdminNotification(
        id: json['id'] as String,
        type: json['type'] as String? ?? '',
        title: json['title'] as String? ?? '',
        message: json['message'] as String? ?? '',
        entityType: json['entityType'] as String?,
        entityId: json['entityId'] as String?,
        isRead: json['isRead'] == true,
        createdAtUtc: DateTime.parse(json['createdAtUtc'] as String).toUtc(),
      );
  AdminNotification asRead() => AdminNotification(
    id: id,
    type: type,
    title: title,
    message: message,
    entityType: entityType,
    entityId: entityId,
    isRead: true,
    createdAtUtc: createdAtUtc,
  );
}

class AdminNotificationPage {
  const AdminNotificationPage(this.items, this.totalCount);
  final List<AdminNotification> items;
  final int totalCount;
}

class NotificationService {
  NotificationService(this._api);
  final ApiClient _api;
  Future<AdminNotificationPage> list(
    String token, {
    int page = 1,
    int pageSize = 20,
    bool unreadOnly = false,
  }) async {
    final json = await _api.getJson(
      '/api/notifications?page=$page&pageSize=$pageSize&unreadOnly=$unreadOnly',
      token: token,
    );
    final items = (json['items'] as List<dynamic>? ?? const [])
        .map((x) => AdminNotification.fromJson(x as Map<String, dynamic>))
        .toList();
    return AdminNotificationPage(
      items,
      json['totalCount'] as int? ?? items.length,
    );
  }

  Future<int> unreadCount(String token) async =>
      (await _api.getJson(
            '/api/notifications/unread-count',
            token: token,
          ))['count']
          as int? ??
      0;
  Future<void> markRead(String token, String id) async {
    await _api.putJson('/api/notifications/$id/read', const {}, token: token);
  }

  Future<void> markAllRead(String token) async {
    await _api.putJson('/api/notifications/read-all', const {}, token: token);
  }
}
